// @katzilla/sdk/dist/client.js
/**
 * Katzilla SDK — HTTP client that talks to api.katzilla.dev.
 *
 * Usage:
 *   const kz = new Katzilla({ apiKey: "kz_abc123" });
 *   const payment = await kz.stripe.createPaymentIntent({ amount: 2000 });
 */
import { StripeAgent } from "./agents/stripe.js";
import { SendGridAgent } from "./agents/sendgrid.js";
import { SupabaseAgent } from "./agents/supabase.js";
import { S3Agent } from "./agents/s3.js";
import { GitHubAgent } from "./agents/github.js";
import { SlackAgent } from "./agents/slack.js";
import { NotionAgent } from "./agents/notion.js";
import { SheetsAgent } from "./agents/sheets.js";
import { CloudflareAgent } from "./agents/cloudflare.js";
import { OpenAIAgent } from "./agents/openai.js";
import { PostHogAgent } from "./agents/posthog.js";
import { SentryAgent } from "./agents/sentry.js";
import { TwilioAgent } from "./agents/twilio.js";
import { UptimeAgent } from "./agents/uptime.js";
export class KatzillaApiError extends Error {
    statusCode;
    body;
    constructor(statusCode, body) {
        super(body.error);
        this.statusCode = statusCode;
        this.body = body;
        this.name = "KatzillaApiError";
    }
}
export class Katzilla {
    apiKey;
    baseUrl;
    // ── Typed agent sub-clients ──────────────────────────────────
    stripe;
    sendgrid;
    supabase;
    s3;
    github;
    slack;
    notion;
    sheets;
    cloudflare;
    openai;
    posthog;
    sentry;
    twilio;
    uptime;
    constructor(options) {
        this.apiKey = options.apiKey;
        this.baseUrl = (options.baseUrl || "https://api.katzilla.dev").replace(/\/$/, "");
        this.stripe = new StripeAgent(this);
        this.sendgrid = new SendGridAgent(this);
        this.supabase = new SupabaseAgent(this);
        this.s3 = new S3Agent(this);
        this.github = new GitHubAgent(this);
        this.slack = new SlackAgent(this);
        this.notion = new NotionAgent(this);
        this.sheets = new SheetsAgent(this);
        this.cloudflare = new CloudflareAgent(this);
        this.openai = new OpenAIAgent(this);
        this.posthog = new PostHogAgent(this);
        this.sentry = new SentryAgent(this);
        this.twilio = new TwilioAgent(this);
        this.uptime = new UptimeAgent(this);
    }
    /**
     * Execute an agent action (dynamic/untyped).
     *
     *   const result = await kz.agent("stripe").action("create-payment-intent", { amount: 2000 });
     */
    agent(handle) {
        return {
            action: (actionId, input = {}) => this.execute(handle, actionId, input),
        };
    }
    /**
     * Low-level execute method. All typed sub-clients delegate here.
     */
    async execute(agentHandle, actionId, input = {}, credentials) {
        const url = `${this.baseUrl}/agents/${agentHandle}/actions/${actionId}`;
        const body = credentials ? { ...input, credentials } : input;
        const res = await fetch(url, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                "X-API-Key": this.apiKey,
            },
            body: JSON.stringify(body),
        });
        const json = await res.json();
        if (!res.ok) {
            throw new KatzillaApiError(res.status, json);
        }
        return json;
    }
}
